[quant][graphmode] `insert_quant_dequant` pass support shared class types

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32379 [WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes
* #32380 [quant][graphmode][refactor] Better API for fold_convbn
* #32374 [refactor] Adding FoldConvBatchNorm2dHelper
* #32303 [quant][graphmode] Call _jit_pass_dedup_module_ueses in quantize_script
* #32302 [quant] Re-enable fold_convbn in quantize_script
* #32206 [quant] Re-enable test_nested that has different qconfig for shared ClassType
* #32205 [quant][graphmode] Support quantizing shared ClassType with different qconfigs
* #32204 [quant][graphmode] Default to non-inplace in graph mode quantization API
* #32203 Adding QConfigTypePtrMap
* #32202 [refactor] Move some of the helper functions for public use
* **#31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types**

Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

jerryzh168 committed:
[quant][graphmode] `insert_quant_dequant` pass support shared class types

Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f34bf24:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f34bf24cf65004e306ab04747d91eac371c2cb6e).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3997922) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76353339) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76353339))
<details>
<summary>
<code>Dec 18 03:15:33 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Dec 18 03:15:33 Ran 2297 tests in 83.289s 
Dec 18 03:15:33  
Dec 18 03:15:33 FAILED (errors=1, skipped=77, expected failures=1) 
Dec 18 03:15:33  
Dec 18 03:15:33 Generating XML reports... 
Dec 18 03:15:33 Traceback (most recent call last): 
Dec 18 03:15:33   File "test/run_test.py", line 456, in <module> 
Dec 18 03:15:33     main() 
Dec 18 03:15:33   File "test/run_test.py", line 449, in main 
Dec 18 03:15:33     raise RuntimeError(message) 
Dec 18 03:15:33 RuntimeError: test_jit_simple failed! 
Dec 18 03:15:33 + cleanup 
Dec 18 03:15:33 + retcode=1 
Dec 18 03:15:33 + set +x 
Dec 18 03:15:33 =================== sccache compilation log =================== 
Dec 18 03:15:33 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 18 03:15:33 Compile requests                56 
Dec 18 03:15:33 Compile requests executed        0 
Dec 18 03:15:33 Cache hits                       0 
Dec 18 03:15:33 Cache misses                     0 
Dec 18 03:15:33 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -852,6 +864,24 @@ void InsertQuantDeQuantHelper::run(
+  // We only need to register new parameters if the graph has
+  // been quantized before
+  if (values_to_quantize_.count(graph.get())) {
+    for (auto& vs : values_to_quantize_.at(graph.get())) {
```
Having the same loop here and in `quantizeTensors` is not great. Any ideas how it can be refactored to reduce code duplication?

jerryzh168 replied:
yeah, I thought about this before, the only difference is one place uses `register_attribute` and the other uses `setAttr`, maybe I can have a function with a boolean input to distinguish between these two cases

jerryzh168 replied:
or I can check whether the attribute exists in the module or not and then switch between these two cases

ZolotukhinM reviewed (approved):
Looks good! You might consider moving the loop you added and the loop from `quantizeTensor` somewhere to reduce code duplication.

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/script/module.h:
```diff
@@ -341,7 +341,7 @@ struct slot_iterator_impl {
     // the last traversal action advanced beyond the number of slots in the
     // module so continue the iteration in the parent.
-    if (top().i_ >= int64_t(top().module_.num_slots())) {
+    if (top().i_ >= int64_t(top().module_._ivalue()->type()->numAttributes())) {
```
@zdevito @suo when we remove attributes for the ClassType that's shared by multiple module instances, there is an inconsistent state where we have removed these attributes in ClassType, but not in the module instance, that's why this change is needed here. If the module is in a consistent state `module_.num_slots()` will always be equal to `module_._ivalue()->type()->numAttributes()` and this change will have no effect. What do you think?
I think we can either do this or create a new Policy for inconsistent modules.

suo replied:
I think it is a big no-no to have the module object state inconsistent with the type. This change proposes the hide the fact that they are inconsistent by switching around the source of truth for a public API, which seems like it will be confusing later.

Is there a way to do this where we never make the modules inconsistent? Two suggestions I see (granted I don't know that much about what this pass is doing so perhaps these are not useful):
1. Don't delete attributes, just construct a new class type with the attributes deleted. (are we already doing this?)
2. When we remove the attributes from the type, also remove them from the module object.

jerryzh168 replied:
The problem is the ClassType is shared by multiple module instances and we want to delete some attribute from the ClassType(and remove slots from the object). (responding to 2.)Suppose there are two module instances share the ClassType, we can safely delete the attribute and also remove the slot from the first module instance, but after that the second module instance will be in a inconsistent state where it has more slots than attributes.
(responding to 1.) If we create a new ClassType, we need to swap out the original ClassType with the new one, and I think that is unsafe as well, and we'll need to do type remapping which also requires more analysis on where the module instance is used.

This is unsafe in general, but in the context of quantization, we know that we inserted the observers ourselves in the insert_observers pass and here we are removing them so I think it is OK to temporary having inconsistent modules in this case.

suo replied:
Okay, as far as the change to the core, I am okay with this. The contract is still that modules are consistent with their types, and in that case this change is a no-op.

I am a little skeptical about this inconsistency you are describing and how easy it will be to maintain for the quant passes, but I guess that's fundamentally up to you. If you'd like we can chat a bit about it; in particular I don't understand how your proposal doesn't error the second time you try `type()->unsafeRemoveAttribute()`—didn't you already remove it for the first module?

jerryzh168 replied:
Yes, second time I'm replaying the removal of slots by index :)

jerryzh168 replied:
I want to talk more about the details for documentation purpose. 
## Inconsistency
### Type 1(`num_slots` < `numAttributes`)
I think the current core has one type of inconsistency which is `num_slots` can be less then `numAttributes` in the type(https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/module.h#L131-L138), in Module::register_attribute, we first `type()->addOrCheckAttribute` and then do `_ivalue()->setAttr(...)` and `setAttr` which calls `setSlot` will automatically create a new slot if the slot index is greater or equal to the size of the slot list, this means after `type()->addOrCheckAttribute` we have more attributes than slots.
### Type 2(`num_slots` > `numAttributes`)
In this PR we introduce another type of inconsistency, which is `numAttributes` is less than `num_slots`, when we remove attribute of a shared ClassType and the slot is not yet removed for some module instances.

## ModulePolicy
Previously we use `num_slots` as the ground truth, and Type 1 inconsistency is the only inconsistency we see, the inconsistency won't be visible, even we iterate through a half constructed module. However, the new type of inconsistency will be visible when we iterate through using the old ModulePolicy, since we have `num_slots` > `numAttributes`.

That's why I changed the ModulePolicy to use `numAttributes` as ground truth, this means if we iterate through a module that has the Type 1 inconsistency, we will see error, but Type 2 inconsistency will not be visible when we iterate use the new ModulePolicy.

I'm assuming Type 1 inconsistency is only there during module construction and after construction we should either have Type 2 inconsistent(as it appears here) objects or consistent objects.

ZolotukhinM replied:
I also squinted a bit about that new inconsistency, but after reading the code carefully I think it should be fine. My ask, however, would be to try to make this code crystal clear - it should be obvious where the inconsistency starts, where it ends and why we never misuse inconsistent objects (e.g. try to double delete).

jerryzh168 replied:
It's all constraint in the insert_quant_dequant pass

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -697,18 +728,31 @@ void InsertQuantDeQuantHelper::removeObservers(
+    for (auto slot: removed_observer_slots_.at(g)) {
+      module._ivalue()->unsafeRemoveSlot(slot);
+    }
+  }
```
@suo here is what we do

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -630,23 +643,41 @@ class InsertQuantDeQuantHelper {
+                  " Got:",
+                  c10::toString(qscheme));
+    } else {
+      qscheme_for_graph_[g] = qscheme;
```
Since we're comparing with `toAffine(qscheme)`, we should probably store `toAffine(qscheme)` rather than plain `qscheme`. Otherwise the check would fail for, e.g, `c10::kPerTensorSymmetric`.

jerryzh168 replied:
you mean moving the toAffine call in comparison here right?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -697,18 +728,31 @@ void InsertQuantDeQuantHelper::removeObservers(
+
+  // If we have seen this graph before, we'll replay the observer
+  // slots removal using the slot index
+  if (removed_observer_slots_.count(g)) {
```
This logic seems more complicated than it could be, or maybe some names are not descriptive enough.

Essentially, we want to remove observers attribute from the type and from objects of this type (there can be several of them). It looks to me that a simple `type_is_visited` or something like this set would reflect that logic better. I think `removed_observer_slots_` is playing this exact role, but I have to think hard and carefully read the code to verify that, which is not great. Separating cleanups for the type and for the objects might make it a bit easier to comprehend. The type cleanup will probably be still called from the object cleanup, but will be guarded by `if(type_cleanedup_.count(type))` check. The overall structure of the code would only slightly differ from what we have now.

jerryzh168 replied:
Do you mean that we don't cleanup the type in the beginning and only do that after all the slots are removed? I don't think that will work because we need to make sure objects and types are in sync for the slot index to make sense.
e.g.
Type A:
   attr1, attr2
object A:
   slot1, slot2
object B:
   slot1, slot2

if we want to remove both slot1 and slot2, and first we remove slot1 from object A, we find the index is 0 from querying type A, and we don't remove attr1 in type A. then object A and type A are not consistent anymore. when we want to remove slot2 from object A, and the query will return the index of 1 which does not exist anymore.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -362,7 +362,7 @@ struct slot_iterator_impl {
   // otherwise, we have to continue advancing.
   bool valid() const {
-    return top().i_ < int64_t(top().module_.num_slots()) &&
+    return top().i_ < int64_t(top().module_._ivalue()->type()->numAttributes()) &&
```
I suspect you might need to clang-format this.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -630,23 +643,41 @@ class InsertQuantDeQuantHelper {
   void collectObserverNodesAndValueToQuantize(script::Module& module, Value*);
-  void removeObservers(script::Module& module);
-  void removeObservers(script::Module& module, Graph* g);
+  void cleanup(script::Module& module);
```
If you decide to use an abstract name `cleanup`, please add comments describing what exactly this cleanup does.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -630,23 +643,41 @@ class InsertQuantDeQuantHelper {
+    } else {
+      qscheme_for_graph_[g] = qscheme;
+    }
+  }
```
Nit: I think there is usually a blank line between functions.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -697,18 +728,31 @@ void InsertQuantDeQuantHelper::removeObservers(
   }
+
+  // If we have seen this graph before, we'll replay the observer
+  // slots removal using the slot index
```
I couldn't understand this comment :) I understood what the code does after reading it, but the comment itself didn't help.

jerryzh168 replied:
Which part is confusing?
I'll write more detailed comment.

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1264,6 +1269,61 @@ def forward(self, x, w0, w1, w2):
+            # Make sure inserted attributes are the same
+            # This needs to be refactored after we know how to get
+            # all attributes of a given ScriptModule
+            for v in ['input.2', 'weight.2', '12']:
```
Don't we have `m.attributes()`? As it is now, this is fragile and relies on internal names that can change at any point.

jerryzh168 replied:
I think we don't have that. Maybe it's good enough to make sure the types are the same

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -630,23 +643,41 @@ class InsertQuantDeQuantHelper {
   void quantizeTensors(script::Module& module, Graph* g, Value* self);
 
  private:
   // TODO: we don't need to call this for each graph
```
Is this TODO still relevant?

jerryzh168 replied:
probably not, I'll remove it. We are using graph to keep track of sharing here.

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Differential Revision: [D19162366](https://our.internmc.facebook.com/intern/diff/D19162366)

[ghstack-poisoned]
